jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudwatch/logs/insights/QueryActor.kt

```diff
@@ -0,0 +1,189 @@
+                } else {
+                    notifyInfo(message("cloudwatch.logs.query_result_completion_status"), message("cloudwatch.logs.query_result_completion_successful"))
+                }
+            }
+                is MessageLoadQueryResults.LoadInitialQueryResults -> {
+                    loadInitialQueryResults()
+                    val rect = table.getCellRect(0, 0, true)
+                    withContext(edtContext) {
+                        table.scrollRectToVisible(rect)
+                    }
+                }
+            }
+        }
+    }
+
+    protected suspend fun loadAndPopulateResultsTable(loadBlock: suspend() -> List<T>) {
+        try {
+            tableLoading()
+            val items = loadBlock()
+            table.listTableModel.items = items
+            table.emptyText.text = emptyText
+        } catch (e: ResourceNotFoundException) {
+            withContext(edtContext) {
+                table.emptyText.text = notFoundText
+            }
+        } catch (e: Exception) {
+            LOG.error(e) { tableErrorMessage }
+            withContext(edtContext) {
+                table.emptyText.text = tableErrorMessage
+                notifyError(title = tableErrorMessage, project = project)
+            }
+        } finally {
+            tableDoneLoading()
+        }
+    }
+
+    protected open suspend fun loadInitialQueryResults() {
+        throw IllegalStateException("Table does not support loadInitialQueryResults")
+    }
+
+    protected abstract suspend fun loadNext(): List<T>
+
+    private suspend fun tableLoading() = withContext(edtContext) {
+        table.setPaintBusy(true)
+        table.emptyText.text = message("loading_resource.loading")
+    }
+
+    private suspend fun tableDoneLoading() = withContext(edtContext) {
+        table.tableViewModel.fireTableDataChanged()
+        table.setPaintBusy(false)
+    }
+
+    override fun dispose() {
+        channel.close()
+    }
+    companion object {
+        private val LOG = getLogger<QueryActor<*>>()
+    }
+}
+
+class QueryResultsActor(
+    project: Project,
+    client: CloudWatchLogsClient,
+    table: TableView<MutableMap<String, String>>,
+    private val queryId: String
+) : QueryActor<MutableMap<String, String>>(project, client, table) {
+    override val emptyText = message("cloudwatch.logs.no_results_found")
+    override val tableErrorMessage = message("cloudwatch.logs.query_results_table_error")
+    override val notFoundText = message("cloudwatch.logs.no_results_found")
+
+    override suspend fun loadInitialQueryResults() {
+        var request = GetQueryResultsRequest.builder().queryId(queryId).build()
+        var response = client.getQueryResults(request)
+        while (response.results().size == 0) {
+            request = GetQueryResultsRequest.builder().queryId(queryId).build()
+            response = client.getQueryResults(request)
+        }
+        val listOfResults = mutableListOf<MutableMap<String, String>>()
```

Comment:
ditto on building with functions instead of a mutable list

Reply:
https://github.com/aws/aws-toolkit-jetbrains/pull/1979#discussion_r471046601